### PR TITLE
Support to JEP-229 / Fixing compilation on multimodule projects

### PR DESCRIPTION
**Pull request by imonteroperez**
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java [hunk 1]
```diff
@@ -113,12 +114,27 @@ private void copy(Path eslintrc, File pluginFolder) {
         return mconfig;
     }
 
-    private void compile(MavenRunner.Config mavenConfig, File path) throws PomExecutionException, IOException {
+    private void compile(MavenRunner.Config mavenConfig, File path, String parentFolder, String pluginName) throws PomExecutionException, IOException {
+        if (StringUtils.isBlank(parentFolder)) {
+            runner.run(mavenConfig, path, setupCompileResources(path), "clean", "process-test-classes", "-Dmaven.javadoc.skip");
+        } else {
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:120]: Only do this if the checkout is in fact a snapshot. https://github.com/jenkinsci/bom/issues/657#issuecomment-970535269 Otherwise in the normal case (any plugin with MRP and not using a local checkout override) you will be `install`ing artifacts into the local repo on top of whatever would otherwise be downloaded from Artifactory, which is risky. https://issues.apache.org/jira/browse/MINSTALL-126 / https://issues.apache.org/jira/browse/MNG-3655

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java [hunk 2]
```diff
@@ -114,22 +117,51 @@ private void copy(Path eslintrc, File pluginFolder) {
...
             // using "mvn clean install -DskipTests -Dinvoker.skip -Denforcer.skip to skip testing and just compile and install dependencies"
             runner.run(mavenConfig, path.getParentFile(), setupCompileResources(path.getParentFile()), "clean", "install", "-DskipTests", "-Dinvoker.skip", "-Denforcer.skip", "-Dmaven.javadoc.skip", "-am", "-pl", pluginName);
+        } else {
+            runner.run(mavenConfig, path, setupCompileResources(path), "clean", "process-test-classes", "-Dmaven.javadoc.skip");
         }
     }
 
+    /**
+     * Checks if a plugin is a multiparent plugin with JEP-229 enabled (see: https://github.com/jenkinsci/jep/blob/master/jep/229/README.adoc)
+     * and an SNAPSHOT project.version and without local checkout directory overriden.
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:132]: The code does what it should be doing, I think, but the method name and Javadoc do not match.

All you need here is to verify that the checkout is of a snapshot. The main way this might happen is in fact that JEP-229 is in use and the commit we are cloning happens to have been promoted as a release. But there might be other ways building a multimodule project could otherwise be broken in a CI system with an empty local repo:

* a regular MRP plugin but overridden with a local clone of a branch or some other non-tag revision
* an incrementalified (JEP-305), but MRP-based not CD-enabled (JEP-229) plugin, checked out at a non-tagged commit (e.g. to test something in a draft PR in `jenkinsci/bom`)
* a non-incrementalified plugin checked out at an arbitrary commit, if PCT in the future supports timestamped snapshots (BEE-1751 in our internal tracker; not very interesting in @jenkinsci since we would just use Incrementals instead)

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 3]
```diff
@@ -892,14 +899,54 @@ private String getProperty(Model model, String value) {
...
         }
     }
+    
+    /**
+     * Provides the Maven module used for a plugin on a `mvn [...] -pl` operation in the parent path 
+     * @param plugin
+     * @param pluginPath
+     * @return the maven module 
+     * @throws PomExecutionException 
+     * @throws IOException
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java:909]:
```suggestion
```

does not add anything

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 4]
```diff
@@ -892,14 +899,54 @@ private String getProperty(Model model, String value) {
             throw new RuntimeException("UpdateSite.Data instantiation problems", e);
         }
     }
+    
+    /**
+     * Provides the Maven module used for a plugin on a `mvn [...] -pl` operation in the parent path
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java:904]:
```suggestion
     * Provides the Maven module used for a plugin on a {@code mvn [...] -pl} operation in the parent path 
```

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 5]
```diff
@@ -892,14 +899,54 @@ private String getProperty(Model model, String value) {
...
+        if (absolutePath.endsWith(plugin)) {
+            return plugin;
+        }
+        String module = absolutePath.substring(absolutePath.lastIndexOf(File.separatorChar) + 1, absolutePath.length());
+        File parentFile = pluginPath.getParentFile();
+        if (parentFile == null) {
+            return null;
+        }
+        File log = new File(parentFile.getAbsolutePath() + File.separatorChar + "modules.log");
+        runner.run(mavenConfig, parentFile, log,"-Dexpression=project.modules", "-q", "-DforceStdout", "help:evaluate");
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java:922]:
```suggestion
        runner.run(mavenConfig, parentFile, log, "-Dexpression=project.modules", "-q", "-DforceStdout", "help:evaluate");
```

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 6]
```diff
@@ -892,14 +899,54 @@ private String getProperty(Model model, String value) {
...
+        for (String line : org.apache.commons.io.FileUtils.readLines(log)) {
+            if (!StringUtils.startsWith(line.trim(), "<string>")) {
+                continue;
+            }
+            String mvnModule = line.replace("<string>", "").replace("</string>", "").trim();
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java:927]: This could be a regexp match perhaps, but why are you attempting to parse XML this way? Use e.g. DOM4j.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 7]
```diff
@@ -1028,7 +1075,7 @@ private void addSplitPluginDependencies(String thisPlugin, MavenRunner.Config mc
             pom.removeDependency(pluginGroupIds.get(thisPlugin), thisPlugin);
         }
     }
-
+
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java:1078]:
```suggestion

```

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java [hunk 8]
```diff
@@ -113,12 +119,60 @@ private void copy(Path eslintrc, File pluginFolder) {
         return mconfig;
     }
 
-    private void compile(MavenRunner.Config mavenConfig, File path) throws PomExecutionException, IOException {
+    private void compile(MavenRunner.Config mavenConfig, File path, File localCheckoutDir, String parentFolder, String pluginName) throws PomExecutionException, IOException {
+        if (isSnapshotMultiParentPlugin(parentFolder, path, localCheckoutDir)) {
+            // "process-test-classes" not working properly on multi-module plugin. See https://issues.jenkins.io/browse/JENKINS-62658
+            // using "mvn clean install -DskipTests -Dinvoker.skip -Denforcer.skip to skip testing and just compile and install dependencies"
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:125]:
```suggestion
            // installs dependencies into local repository
```

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java [hunk 9]
```diff
@@ -113,12 +119,60 @@ private void copy(Path eslintrc, File pluginFolder) {
...
+     * @param path
+     * @param localCheckoutDir
+     * @return
+     * @throws PomExecutionException 
+     * @throws IOException 
+     */
+    private boolean isSnapshotMultiParentPlugin(String parentFolder, File path, File localCheckoutDir) throws PomExecutionException, IOException {
+        if (localCheckoutDir != null) {
+            return false;
+        }
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:150]: I do not understand why this clause is here. As discussed earlier, having a local checkout dir (with a plain old MRP plugin) is one of the reasons why you _might_ in fact need this patch.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java [hunk 10]
```diff
@@ -113,12 +119,60 @@ private void copy(Path eslintrc, File pluginFolder) {
...
+            }
+            runner.run(mavenConfig, path.getParentFile(), setupCompileResources(path.getParentFile()), "clean", "install", "-DskipTests", "-Dinvoker.skip", "-Denforcer.skip", "-Dmaven.javadoc.skip", "-am", "-pl", mavenModule);
+        } else {
+            runner.run(mavenConfig, path, setupCompileResources(path), "clean", "process-test-classes", "-Dmaven.javadoc.skip");
+        }
+    }
+
+    /**
+     * Checks if a plugin is a multiparent plugin with a SNAPSHOT project.version and 
+     * without local checkout directory overriden.
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:138]: ditto

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java [hunk 11]
```diff
@@ -113,12 +119,60 @@ private void copy(Path eslintrc, File pluginFolder) {
...
+    /**
+     * Checks if a plugin is a multiparent plugin with a SNAPSHOT project.version and 
+     * without local checkout directory overriden. 
+     * 
+     * @param parentFolder
+     * @param path
+     * @param localCheckoutDir
+     * @return
+     * @throws PomExecutionException 
+     * @throws IOException
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:145]:
```suggestion
```

ditto

**Review by jglick — APPROVED** [review]
Looks OK. Does it in fact work?

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 6 above)
Comment (imonteroperez) [reply plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java:922]: Yep, I will update, trying a quick-and-dirty approach just to test the entire solution

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java (hunk 9 above)
Comment (imonteroperez) [reply plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:143]: :thinking: you are right, I will review this in depth. I included at the very beginning because I wrongly understood https://github.com/jenkinsci/plugin-compat-tester/pull/318#discussion_r752494751

Comment (jglick) [reply plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/MultiParentCompileHook.java:143]: Definitely an edge case, not very important, just a point of confusion.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 6 above)
Comment (jglick) [reply plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java:922]: Yes. Good enough for now I think.

**Comment by imonteroperez** [comment]
> Looks OK. Does it in fact work?

Yes! it worked fine, before this PR
```
[ERROR] Failed to execute goal on project aws-java-sdk-ec2: Could not resolve dependencies for project org.jenkins-ci.plugins.aws-java-sdk:aws-java-sdk-ec2:hpi:1.12.101-999999-SNAPSHOT: Could not find artifact org.jenkins-ci.plugins.aws-java-sdk:aws-java-sdk-minimal:jar:1.12.101-999999-SNAPSHOT
```
and after this PR
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Amazon Web Services SDK :: Parent 1.12.101-999999-SNAPSHOT:
[INFO] 
[INFO] Amazon Web Services SDK :: Parent .................. SUCCESS [  0.477 s]
[INFO] Amazon Web Services SDK :: Minimal ................. SUCCESS [  0.130 s]
[INFO] Amazon Web Services SDK :: EC2 ..................... SUCCESS [  0.065 s]
[INFO] ------------------------------------------------------------------------
[...]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running InjectedTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 8.335 s - in InjectedTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
```

**Review by cathychan — APPROVED** [review]
LGTM!

**Comment by jglick** [comment]
@imonteroperez you need to set a content label for CD before merging a PR.
